test/unit/org/apache/cassandra/io/sstable/SSTableReaderTest.java

```diff
@@ -140,6 +150,80 @@ public void testGetPositionsForRanges()
+        if (partitioner.splitter().isPresent())
+            tokens.add(partitioner.getMaximumToken());
+
+        for (int j = 0; j < 100; j++)
+        {
+            Mutation mutation = new RowUpdateBuilder(store.metadata(), j, String.valueOf(random.nextInt())).clustering("0")
+                                                                                                           .add("val",
+                                                                                                                ByteBufferUtil.EMPTY_BYTE_BUFFER)
+                                                                                                           .build();
+            if (j % 4 != 0) // skip some keys
+                mutation.applyUnsafe();
+            tokens.add(mutation.key().getToken());
+        }
+
+        store.forceBlockingFlush();
+        assertEquals(1, store.getLiveSSTables().size());
+        SSTableReader sstable = store.getLiveSSTables().iterator().next();
+
+        // verify any combination of start and end point among the keys we have, which includes empty, full and
+        // wrap-around ranges
+        for (int i = 0; i < tokens.size(); i++)
+            for (int j = 0; j < tokens.size(); j++)
+            {
+                verifyEstimatedKeysAndKeySamples(sstable, new Range<Token>(tokens.get(i), tokens.get(j)));
+            }
+    }
+
+    private void verifyEstimatedKeysAndKeySamples(SSTableReader sstable, Range<Token> range)
+    {
+        List<DecoratedKey> expectedKeys = new ArrayList<>();
+        try (ISSTableScanner scanner = sstable.getScanner())
+        {
+            while (scanner.hasNext())
+            {
+                try (UnfilteredRowIterator rowIterator = scanner.next())
+                {
+                    if (range.contains(rowIterator.partitionKey().getToken()))
+                        expectedKeys.add(rowIterator.partitionKey());
+                }
+            }
+        }
+
+        // check estimated key
+        long estimated = sstable.estimatedKeysForRanges(Collections.singleton(range));
+        assertTrue("Range: " + range + " having " + expectedKeys.size() + " partitions, but estimated "
+                   + estimated, closeEstimation(expectedKeys.size(), estimated));
+
+        // check key samples
+        List<DecoratedKey> sampledKeys = new ArrayList<>();
+        sstable.getKeySamples(range).forEach(sampledKeys::add);
+
+        assertTrue("Range: " + range + " having " + expectedKeys + " keys, but keys sampled: "
+                   + sampledKeys, expectedKeys.containsAll(sampledKeys));
+        // no duplicate
+        assertEquals(expectedKeys.size(), expectedKeys.stream().distinct().count());
+        assertEquals(sampledKeys.size(), sampledKeys.stream().distinct().count());
+    }
+
+    private boolean closeEstimation(long expected, long estimated)
+    {
+        return expected <= estimated + 16 && expected >= estimated - 16;
```

Comment:
Is there any specific reason why the estimation error is the magic number 16?

Reply:
Well... in this environment (OSS C* with `big` sstable format) it's 16 because we set up the index summary to be on every 8 (i.e. up to 8 error on either side --> 16 max). In dse and `ds-trunk`, it's 16 because the approximation the `bti` sstable format does (which is based on average partition size and is not as stable) comes out to below 16 difference for this test...

I.e. there is a good reason why it's 16 now, but that's not the reason 16 is still a good choice once we merge this into stargazer. Which means we have the choice 
- to add a comment that will be incorrect (and not corrected) in our main branch
- to add a comment describing features not yet in C*
- to add a vague comment that doesn't point out current reason
- to leave it without comment

Personally, I still think the latter is the best choice (short of purposefully breaking the test for `ds-trunk`), but let me know if you disagree and insist on a different one.

Reply:
Thank you for explaining. So this magic number is not specific to this test and is associated with the actual sstable format, right? IMHO, it is good to have such semantics visible in the code. However, I am not sure it is inside the scope of the PR. And I don't think a comment is a good way to bring the semantics.